Refactor and fix the behavior of debuginfo-install

lukash opened the pull request:
Basically rewrites the plugin to:
* Install the reuqested version if specified on the command line

* Install the version of the installed base package if it's installed

* Install the latest available debuginfo version of packages otherwise

* Properly install both debuginfo and debugsource packages where
available

* Print informative messages for packages that do not have debuginfo

* More effectively query and install the packages in one step

https://bugzilla.redhat.com/show_bug.cgi?id=1586059
https://bugzilla.redhat.com/show_bug.cgi?id=1586084

---

Requires this: https://github.com/rpm-software-management/dnf/pull/1425
The tests are here: https://github.com/rpm-software-management/ci-dnf-stack/pull/566

lukash commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+                        packages.pop(na)
+                if na[0].endswith(Package.DEBUGSOURCE_SUFFIX):
+                    if (na[0][:-len(Package.DEBUGSOURCE_SUFFIX)], na[1]) in packages:
+                        packages.pop(na)
```
This is super-ugly. Help!

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+            # installed version of the base package, as that would require
+            # another query and is further complicated if the user specifies a
+            # version himself etc.
+            for na in list(packages.keys()):
```
I would recommend to use a query instead with query = query.union(query2) or difference or so on functions.

lukash replied:
How exactly would you use it in this case? I am doing a check for each package I'm iterating through. An alternative could be to make a list of packages that match the condition and filter those out - but the list can be potentially very long, and it will be a list of (name, architecture) pairs. Is this what you mean?

pkratoch replied:
So, the root cause why you are doing it is that ``debuginfo-install foo-debuginfo`` installs the newest version and does not match the version of ``foo``. And in the specific case of ``debuginfo-install foo foo-debuginfo`` (or similarly with globs), this fixes the problem. Right?

I get that you cannot just naively check the name without the suffix... but you are actually doing that here. So why not just check the existence of the base package, not just its presence in the packages? Would it be too slow? I really don't know, so sorry if you already thought about this.

I am thinking about doing one query for all the base names, installing these and then just checking that all the debuginfo names are in the packages_installed; and if not, installing them, too. Also, I don't know what @j-mracek had in mind, so maybe he has some good idea.

lukash replied:
> So, the root cause why you are doing it is that `debuginfo-install foo-debuginfo` installs the newest version and does not match the version of `foo`. And in the specific case of `debuginfo-install foo foo-debuginfo` (or similarly with globs), this fixes the problem. Right?

Yes.

> I get that you cannot just naively check the name without the suffix... but you are actually doing that here. So why not just check the existence of the base package, not just its presence in the packages? Would it be too slow? I really don't know, so sorry if you already thought about this.
> 
> I am thinking about doing one query for all the base names, installing these and then just checking that all the debuginfo names are in the packages_installed; and if not, installing them, too. Also, I don't know what @j-mracek had in mind, so maybe he has some good idea.

I'm not sure how this would be better? I still need to separate the packages to "base" and "debuginfo" by checking for the suffix, and then I'd have to do this in two steps, which I think would also result in not installing a newer version of e.g. `foo-debuginfo` with `dnf debuginfo-install foo-debuginfo` if an older version of `foo-debuginfo` was already installed? But maybe I'm just not getting it...

If you are thinking of performance overhead of using the NA (name, architecture) dicts, I think it's not a concern, I need to use the dicts anyway (to squash the installed and available package lists) and then the lookup is just O(1). The number of packages also shouldn't be huge.

lukash commented:
I've added dependency on dnf-4.2.9 to the spec file.

pkratoch commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+                "release" not in kwargs:
+            kwargs["epoch"] = package.epoch
+            kwargs["version"] = package.version
+            kwargs["release"] = package.release
```
When I have ``foo-1.0-1`` installed and ``foo-1.0-2`` available, then ``debuginfo-install foo-1.0`` installs ``foo-1.0-2``. Can it also match the installed package if the provided attributes are the same as of the installed package? That behavior seems better to me. For example like this:
```
        if package._from_system and \
                ("epoch" not in kwargs or kwargs["epoch"] == package.epoch) and \
                ("version" not in kwargs or kwargs["version"] == package.version) and \
                ("release" not in kwargs or kwargs["release"] == package.release):
            kwargs["epoch"] = package.epoch
            kwargs["version"] = package.version
            kwargs["release"] = package.release
```

lukash replied:
Good idea, I'll do it.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+            # another query and is further complicated if the user specifies a
+            # version himself etc.
+            for na in list(packages.keys()):
+                if na[0].endswith(Package.DEBUGINFO_SUFFIX):
```
Providing a constant by Package.DEBUGINFO_SUFFIX is different way that was used so far for name of system repo or so on.

lukash replied:
I'm not sure what you mean :confused:

j-mracek replied:
What about to put DEBUGINFO_SUFFIX outside of package class

lukash replied:
Why?

j-mracek replied:
Debuginfo surfix is identical for package and for repo.

j-mracek replied:
Anyway add local variable debuginfo_surfix = Package.DEBUGINFO_SUFFIX. I believe that it is a better to not call package object to get static value.

lukash replied:
Hmm, sorry, still not entirely clear.. so in stead of:

```python
class Package:
    DEBUGINFO_SUFFIX = "-debuginfo"
    # ...
```

You want me to put the constant straight into the module like:
```python
DEBUGINFO_SUFFIX = "-debuginfo"

class Package:
    # ...
```

At least that's what I think makes sense and I can do that... But it doesn't seem exactly what you're suggesting?

j-mracek replied:
Yes, that what I suggested. The question is in which module in libdnf.

lukash replied:
I clearly need it for `package` here. IIUC you're saying the suffix is the same for repositories, but that is sort of a coincidence and it's also just a convention? We shouldn't try to link the two I think. I'd put it in `package`. I kind of liked it in the class, since that's the one working with the suffixes and it kinda feels it belongs there...

j-mracek replied:
Please make change according to your best opinion

lukash replied:
After some discussion with @m-blaha I'll keep the constants inside the class.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+                    if (na[0][:-len(Package.DEBUGSOURCE_SUFFIX)], na[1]) in packages:
+                        packages.pop(na)
+
+            if not packages:
```
The if could be moved several lines up and could end with "continue" line, because there is no packages.key() and no package.values(). Pop up statement still keep non-debuginfo packages in the dict.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+                    if (na[0][:-len(Package.DEBUGINFO_SUFFIX)], na[1]) in packages:
+                        packages.pop(na)
+                if na[0].endswith(Package.DEBUGSOURCE_SUFFIX):
+                    if (na[0][:-len(Package.DEBUGSOURCE_SUFFIX)], na[1]) in packages:
```
The len of suffix could be calculated once.

lukash replied:
The complexity of len() in python is O(1), there is no calculation being done, but I wonder if the interpreter will optimize out the function calls or calls the function again and again multiple times. Brief research shows it does some optimizations of constants (which this is), but there is a function call so I'm not sure it can figure out it is a constant. Unless someone knows better if it does optimize, I'll store the lengths in variables.

j-mracek replied:
Perfect

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+        # if the package is installed and none of EVR was specified in the
+        # argument, use the installed package's EVR (install debuginfo with the
+        # same version as the installed package)
+        if package._from_system and \
```
I believe there is a problem with arch value in case that secondary architecture is installed.
foo.i686 installed
I believe that with argument foo, debuginfo for x86_64 will be installed. I am proposing to add:
```
if package._from_system:
    kwargs["arch"] = package.arch
```

j-mracek replied:
Also globs in version, release, arch will make the function non-functional.
Keuwords could be replaced like "release__glob".
I am also not sure, how the if will handle it.
Probably conditions ```                 ("epoch" not in kwargs or kwargs["epoch"] == package.epoch) and \
                ("version" not in kwargs or kwargs["version"] == package.version) and \
                ("release" not in kwargs or kwargs["release"] == package.release): ```
could be removed, because installed package already matched version, release or so on in previous query.

lukash replied:
> Also globs in version, release, arch will make the function non-functional.
> Keuwords could be replaced like "release__glob".
> I am also not sure, how the if will handle it.

Good observation. I've just tested it, it will not find the correct package when a glob is specified. However, what is a glob supposed to do here? The way it's supposed to work is the user specifies a version of the `-debuginfo` to install. If they specify a glob, which version should it install? Like the highest one matching the glob? I'm not sure we want to support this at all?

> Probably conditions ("epoch" not in kwargs or kwargs["epoch"] == package.epoch) and \ ("version" not in kwargs or kwargs["version"] == package.version) and \ ("release" not in kwargs or kwargs["release"] == package.release):
could be removed, because installed package already matched version, release or so on in previous query.

You're right, I think it can actually be removed and the tests do pass with it gone... :slightly_smiling_face:

j-mracek replied:
When glob is used in version i would suggest:
1. In set is installed package use version of installed package
2. Let solver to decide - mostly the highest version
You only need to change kwargs["version"] to kwargs["version__glob"], kwargs["release"] to kwargs["release__glob"], and kwargs["arch"] to kwargs["arch__glob"]. parsing epoch do not support globs.
The rest of the code will handle it correctly.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+        query = self.packages_available.filter(name=debug_name, **kwargs)
+
+        if query:
+            s = dnf.selector.Selector(self.base.sack)
```
Please could you rename s to sltr or something else? It will look better.

lukash replied:
For this three-line scope and clear definition of what `s` is on the right side of `=` I think one letter variables are fine, but I've renamed it to `selector`.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+            # attempt to install debuginfo and debugsource for the highest
+            # listed version of the package (in case the package is installed,
+            # only the installed version is listed)
+            for pkg in packages.values():
```
I would recommend to add support when 'foo-debuginfo' argument is used. Debuginfo command not only search for packages, but also enebles debuginfo repositories. What about the change like this?
```
            for pkgs in packages.values():
                package = pkgs[0]
                pkg_name = package.name
                if pkg_name.endswith(Package.DEBUGINFO_SUFFIX) or pkg_name.endswith(Package.DEBUGSOURCE_SUFFIX):
                    sltr = dnf.selector.Selector(self.base.sack)
                    sltr.set(pkg=pkgs)
                    self.base.goal.install(select=sltr, optional=not self.base.conf.strict)
                else:
                    if not self._install(package, package.debug_name, solution["nevra"]):
                        self.debuginfo_missing.add(str(package))

                    if not self._install(package, package.debugsource_name, solution["nevra"]):
                        self.debugsource_missing.add(str(package))
```

pkratoch replied:
There is a support for ``foo-debuginfo`` arguments. Note that Lukas changed/added the ``debug_name`` and ``debugsource_name`` functions in https://github.com/rpm-software-management/dnf/pull/1425 so that it handles the suffixes better.

Debuginfo repositories are enabled in ``self.configure()``. Or what did you mean?

j-mracek replied:
Thanks Pavla I didn't know that. But anyway the proposed change will spare another query, because the search result should be identical.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
-            logger.info(
-                _("Could not find debuginfo for package: %s") % package)
+    def _install(self, package, debug_name, base_nevra):
+        if debug_name in self.packages_installed:
```
Is this condition really necessary? When I use argument foo-1 and foo-2 and both version will exist, the code will install packages for the first argument. The second argument will be silently skipped. By removal of that condition it will fail with depsolver error.

lukash replied:
This is one of the few remnants of the original code :slightly_smiling_face: It's clearly supposed to prevent reinstalling the same packages many times over. However, the `self.base.goal` already does that (and better), so I'll drop this.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -79,54 +83,95 @@ def run(self):
+        kwargs = {}
+
+        # if some part of EVRA was specified in the argument, add it as a filter
+        if base_nevra.epoch is not None:
```
base_nevra could be None. Please try argument like "/etc/dnf/dnf.conf", or provide, or NEVRA when "-" or ":" will be replaced by "*"

j-mracek commented:
Please reference also this bug - https://bugzilla.redhat.com/show_bug.cgi?id=1532378

lukash commented:
The resolved comments were addressed in the latest commit.

The secondary architecture use-case and the non-existent nevra will require big changes to make them work though.

lukash commented:
I've pushed a newer version that deals with different architectures of installed packages. I'm now using `query._name_dict()` instead of `query._na_dict()`, as that didn't work for the multiple architecture case. But the `_name_dict()` still has a problem, because the package lists in its values can contain packages with different achitectures and sometimes the wrong architecture is first in the list, and it can:
1. be logged in case `-debuginfo` wasn't found, confusing the user
2. in case NEVRA is None (as @j-mracek notes above), the only information we have is this package, and it is wrong

I've written tests for these two cases here, they're marked with @xfail: https://github.com/rpm-software-management/ci-dnf-stack/pull/566

The problems are also marked with TODOs in the current code.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +75,126 @@ def configure(self):
+                                               epoch=package.epoch,
+                                               version=package.version,
+                                               release=package.release,
+                                               arch=package.arch)
```
Please replace self.packages_available by self.base.sack.query() to prevent reinstalling of packages

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +75,126 @@ def configure(self):
+        if base_nevra.arch is not None:
+            kwargs["arch__glob"] = base_nevra.arch
+
+        query = self.packages_available.filter(name=debug_name, **kwargs)
```
Please could you replace self.packages_available by self.base.query to prevent reinstalling installed packages

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +75,125 @@ def configure(self):
+            # version themselves etc.
+            for name in list(package_dict.keys()):
+                if name.endswith(Package.DEBUGINFO_SUFFIX):
+                    if name[:-debuginfo_suffix_len] in package_dict:
```
Please add condition for testing arch. This could be very tricky because foo-debuginfo could be in dist with a single architecture and foo could have also a single or multiple arch packages in dict.

lukash replied:
Not sure I understand you, but if you mean I should test if the `-debuginfo` package in the dict has the same architecture as the base package I'm looking up here, I don't think it's needed.

I don't think we really care about the architecture now, we are just detecting a case when both `-debuginfo` and base package matched a single input pkgspec. And if it did, we want to remove the `-debuginfo` package, becase it is superfluous here and breaks the correct handling of installing the debuginfo for the base package later, which will take care of the architectures too...

It's a bit complicated but I think there is no problem here, in case you still think it is, can you think of a test case I could add for it?

j-mracek replied:
Probably you all right here.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +75,125 @@ def configure(self):
+                    # for packages from system (installed) there can be more
+                    # packages with different architectures listed and we want
+                    # to install debuginfo for all of them
+                    for package in packages:
```
I would recommend a different approach. It is little bit complicated therefore I provide a diff
```
diff --git a/plugins/debuginfo-install.py b/plugins/debuginfo-install.py
index 4b53565..28739d1 100644
--- a/plugins/debuginfo-install.py
+++ b/plugins/debuginfo-install.py
@@ -116,14 +116,18 @@ class DebuginfoInstallCommand(dnf.cli.Command):
             # listed version of the package (in case the package is installed,
             # only the installed version is listed)
             for packages in package_dict.values():
+                arch_dict = {}
+
                 package = packages[0]
 
-                if packages[0]._from_system:
+                if package._from_system:
                     # for packages from system (installed) there can be more
                     # packages with different architectures listed and we want
                     # to install debuginfo for all of them
-                    for package in packages:
-                        if not self._install_debug_from_system(package, package.debug_name):
+                    for pkg in packages:
+                        arch_dict.setdefault(pkg.arch, []).append(pkg)
+                    for pkgs_per_arch in arch_dict.values():
+                        if not self._install_debug_from_system(pkgs_per_arch, package.debug_name):
                             self.debuginfo_missing.add(str(package))
 
                         if not self._install_debug_from_system(package, package.debugsource_name):
@@ -131,7 +135,6 @@ class DebuginfoInstallCommand(dnf.cli.Command):
 
                 else:
                     # if the first package is not from the system, we just want its name
-                    package = packages[0]
                     if package.name.endswith(Package.DEBUGINFO_SUFFIX) \
                             or package.name.endswith(Package.DEBUGSOURCE_SUFFIX):
 
@@ -159,12 +162,14 @@ class DebuginfoInstallCommand(dnf.cli.Command):
             raise dnf.exceptions.PackagesNotAvailableError(_("Unable to find a match"),
                                                            pkg_spec=' '.join(errors_spec))
 
-    def _install_debug_from_system(self, package, debug_name):
-        query = self.base.sack.query().filter(name=debug_name,
-                                              epoch=package.epoch,
-                                              version=package.version,
-                                              release=package.release,
-                                              arch=package.arch)
+    def _install_debug_from_system(self, per_arch_packages, debug_name):
+        base_query = self.base.sack.query().filterm(arch=per_arch_packages[0].arch).apply()
+        query = self.base.sack.query().filterm(empty=True)
+        for package in per_arch_packages:
+            query.union(self.base.sack.query().filter(name=debug_name,
+                                                      epoch=package.epoch,
+                                                      version=package.version,
+                                                      release=package.release))
 
         if query:
             self._install(query)
(END)

```

j-mracek replied:
Also part
```
        base_query = self.base.sack.query().filterm(arch=per_arch_packages[0].arch).apply()
        query = self.base.sack.query().filterm(empty=True)
        for package in per_arch_packages:
            query.union(base_query.filter(name=debug_name,
                                          epoch=package.epoch,
                                          version=package.version,
                                          release=package.release))
```
could be replaced by:
```
nevras=[]
        for pkg in per_arch_packages:
            nevras.append("{}-{}:{}-{}.{}".format(
                debug_name, pkg.epoch, pkg.version, pkg.release, pkg.arch))
        query = self.base.sack.query().filterm(nevra_strict=nevras)
```
The second option would be more efficient.

lukash replied:
@j-mracek first, I'd really like to stress that I _need_ some sort of description of your proposed solutions. Most of the time I have to seriously try hard to figure out what you actually mean...

It seems you've got some sort of error in your patch, the `base_query` you create in `_install_debug_from_system` is never used. Even with that fixed though, I don't understand how your solution practically differs from what my code is doing...

And, on top of that, this doesn't address neither of the two problems I've described above :slightly_smiling_face::


> <snip> But the `_name_dict()` still has a problem, because the package lists in its values can contain packages with different achitectures and sometimes the wrong architecture is first in the list, and it can:
> 
> 1. be logged in case `-debuginfo` wasn't found, confusing the user
> 
> 2. in case NEVRA is None (as @j-mracek notes above), the only information we have is this package, and it is wrong

Please do see the tests in the CI repo, they demonstrate well what doesn't work. The two tests in question are marked with @xfail: [rpm-software-management/ci-dnf-stack#566](https://github.com/rpm-software-management/ci-dnf-stack/pull/566)

lukash replied:
So the problem actually was with installonly packages. I've added a test to the tests PR and I've fixed the code to install only the latest version for each architecture, based on the above suggestion, but I've simplified it a bit, installing only the latest package for each architecture.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +75,125 @@ def configure(self):
+                        print("TODO We have this package and it may be the wrong one: %s", package)
+
+                    if not self._install_debug(package.debug_name, solution["nevra"]):
+                        self.debuginfo_missing.add(str(package)) # TODO check this is the correct package being logged - could be different architecture etc.
```
To report latest packages for all arch it can be resolve by:
```
    def _latest_arch_representant(self, packages):
        """Packages should be sorted list"""
        latest_per_arch = [packages[0]]
        arch = packages[0].arch
        for pkg in packages[1:]:
            if arch != pkg.arch:
                latest_per_arch.append(pkg)
                arch = pkg.arch
        return latest_per_arch
```
and self.debuginfo_missing.add(str(package)) could be replaced by:
```
for pkg in _latest_arch_representant(packages):
    self.debuginfo_missing.add(str(pkg))
```

lukash replied:
I've split the _Could not find ..._ messages instead and for installed packages I print out the full NEVRA, for available packages I print out NEVR, omitting the architecture, of which there can be multiple and it can be confusing listing the same package multiple times with different architectures.

j-mracek replied:
ok

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +75,133 @@ def configure(self):
+                        continue
+
+                    if solution["nevra"] is None:
+                        print("TODO We have this package and it may be the wrong one: %s", package)
```
What about to use a similar construct like for installed packages?
```
                        if not self._install_debug_no_nevra(packages, package.debug_name):
                            # same problem with logging and relevant arch
                            self.debuginfo_missing.add(str(package))

                        if not self._install_debug_no_nevra(packages, package.debugsource_name):
                            self.debugsource_missing.add(str(package))
```
```
    def _install_debug_no_nevra(self, packages, debug_name):
        """The function is identical with proposed change of _install_debug_from_system"""
        nevras = []
        for pkg in packages:
            nevras.append("{}-{}:{}-{}.{}".format(
                debug_name, pkg.epoch, pkg.version, pkg.release, pkg.arch))
        query = self.base.sack.query().filterm(nevra_strict=nevras)
        if query:
            self._install(query)
            return True

        return False
```

lukash replied:
I've used this approach, just simplified a bit the `nevras` composition.

lukash commented:
Pushed a new version which I hope is more or less final (apart from perhaps moving the `DEBUGINFO_SUFFIX` and `DEBUGSOURCE_SUFFIX` constants?). I've made some attempt to restructure the code a bit better as well.

@j-mracek please go through the unresolved conversations and see if you still find any upstanding issues (feel free to resolve the ones where you don't).

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -46,8 +45,7 @@ def config(self):
             # allow update of already installed debuginfo packages
-            dbginfo = dnf.sack._rpmdb_sack(self.base).query().filter(
-                                                name__glob="*-debuginfo")
+            dbginfo = dnf.sack._rpmdb_sack(self.base).query().filter(name__glob="*-debuginfo")
```
Please replace `filter` by `filterm`. `filter` requrns new query, `filterm` modifies the present query.

lukash replied:
Why? Note I just changed the formatting here sort of ad-hoc... I'm not entirely sure what the code does, but it creates a `dbginfo` variable and stores the new query in there. So I don't see how using `filterm` would be correct here.

m-blaha replied:
You create new query by dnf.sack._rpmdb_sack(self.base).query(), then .filter() method creates and returns new query based on the original one. filterm() does the modification in-place. You would spare one query creation.

lukash replied:
Right, I didn't realize `filterm()` also returns the query. Thanks.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
-            if not package_query:
-                msg = _('No match for argument: %s')
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
```
Please add one key word:
`solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack, with_src=False)`
In some cases when glob is used in combination with parser it ciuld result that it returns only src packages and line below will remove them. I don't remember any example.

lukash replied:
Will do.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
+
+            query = solution["query"].filterm(arch__neq='src')
```
Can be removed, see above.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+            # installed versions of packages have priority, replace / add them to the dict
+            package_dict.update(query.installed()._name_dict())
+
+            if not package_dict:
```
Please could you move the code 
```

            if not query:
                logger.info(_('No match for argument: %s'), self.base.output.term.bold(pkgspec))
                errors_spec.append(pkgspec)
                continue
```
before creation of dicts. It is more efficient and transparent.

lukash replied:
Will do.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+                if first_package.name.endswith(Package.DEBUGINFO_SUFFIX) \
+                        or first_package.name.endswith(Package.DEBUGSOURCE_SUFFIX):
+
+                    self._install((first_package, ))
```
I would recommend to use all packages that we got like self._install(packages). To keep more options to solver is beneficial, because user could use arguments that have only limited overlap. Like `foo-debuginfo` and `foo-debuginfo-1-5`. Using all packages from the query is not harmful.

j-mracek replied:
@lukash Please could you consider the proposed change?

lukash replied:
Done.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+
+    def _install_debug_no_nevra(self, debug_name, packages):
+        query = self.base.sack.query().filterm(
+            nevra_strict=["{}-{}.{}".format(debug_name, p.evr, p.arch) for p in packages])
```
Please could you rename p to pkg. This is more common name in the project

lukash replied:
I would actually rather prefer a single-letter variable here. The scope is limited to this single line, so there is no doubt what it represents and the single letter makes the line shorter and easier to read. I think single letters are common in constructs like list comprehension and good to use there...

lukash replied:
I could, though, rename:
`packages -> pkgs`
`package -> pkg`

Throughout the code. I does save quite a few characters overall and if it's sort of a convention...

j-mracek replied:
Proposed change `packages -> pkgs` `package -> pkg` is fine.

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+
+                    continue
+
+                # inf we don't have NEVRA from the pkgspec, pass nevras from
```
Typo if not inf

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+
+                    continue
+
+                # if the package in question is -debuginfo or -debugsource, install it directly
```
question? Did you mean query?

lukash replied:
No, it's a phrase, meaning the package we are working with. I think it's fine to use it here...

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+                        package_dict.pop(name)
+                if name.endswith(Package.DEBUGSOURCE_SUFFIX):
+                    if name[:-debugsource_suffix_len] in package_dict:
+                        package_dict.pop(name)
```
I believe that after the latest changes `for name in list(package_dict.keys()):` loop could be removed completely. If we will provide all possibility to solver (list of all packages and not only the latest) there is no need to reduce data set.

lukash replied:
This won't work for the case when there is e.g. package `bar-2.0` installed and `bar-{debuginfo,debugsource}-2.0` don't exist. Expected behavior is no `bar-{debuginfo,debugsource}` aren't installed, but if I remove this piece of code, it will install the latest versions of `bar-{debuginfo,debugsource}` available.

There is a testcase with this scenario (called _"multiple packages, a glob that doesn't find debuginfo for the installed version"_) in the CI PR, which breaks if I try this (even if I pass all the `-debuginfo` packages to the solver instead of just the first one).

j-mracek replied:
Then it must be resolved somewhere else.

j-mracek replied:
> This won't work for the case when there is e.g. package `bar-2.0` installed and `bar-{debuginfo,debugsource}-2.0` don't exist. Expected behavior is no `bar-{debuginfo,debugsource}` are installed, but if I remove this piece of code, it will install the latest versions of `bar-{debuginfo,debugsource}` available.

In this case I believe that it should fail.

> 
> There is a testcase with this scenario (called _"multiple packages, a glob that doesn't find debuginfo for the installed version"_) in the CI PR, which breaks if I try this (even if I pass all the `-debuginfo` packages to the solver instead of just the first one).

j-mracek replied:
Please provide detailed example including set packages installed, available and argument used for command.

lukash replied:
The referenced test case is just that...

j-mracek commented on plugins/debuginfo-install.py:
```diff
@@ -74,59 +76,170 @@ def configure(self):
+                _("Could not find debugsource package for the following installed packages: %s"),
+                ", ".join(sorted(self.installed_debugsource_missing)))
 
         if errors_spec and self.base.conf.strict:
```
Why missing installed_debugsource_missing does not trigger an error? Then we can remove one loop above (`for name in list(package_dict.keys()):`).

lukash replied:
I don't understand what you mean...

j-mracek replied:
What about to replace it by:
`if (errors_spec or self.installed_debugsource_missing) and self.base.conf.strict:`

I suggest that this will resolve the issue when debuginfo is present but not in version of installed package and argument like (foo*) is used.

lukash replied:
But how will it resolve the issue? This is just reporting an error at the end, how can it resolve the issue?

Also you've picked one of the four lists of missing packages, so I just don't see how that makes sense?

(I've pushed fixes for the minor comments you had)

pkratoch replied:
@j-mracek,  you have probably missed that based on the discussion in [bug 1586059](https://bugzilla.redhat.com/show_bug.cgi?id=1586059), there should be error only when the pkgspec doesn't match anything. If -debuginfo/-debugsource package is missing, it's not an error, because not all packages must have debuginfo. See [comment 15](https://bugzilla.redhat.com/show_bug.cgi?id=1586059#c15) for a reasoning.

lukash replied:
(I somehow thought the discussion is still about the `for name in list(package_dict.keys()):` "hack" way earlier in the code, also discussed in the thread above. Thanks @pkratoch)

lukash commented:
I've found out that the feature of trying to install `-debuginfo` package from the base (parent) package in case it is not found for a subpackage (e.g. `debuinfo install foo-subpackage` will try to install `foo-debuginfo` if `foo-subpackage-debuginfo` is not found), which I've removed originally, is required, because e.g. `glibc` does have the debuginfo set up this way.

Luckily it wasn't too hard to add, although it does add one more twist to the plot...

I've pushed a fixup commit for now, so that it's clear what I've changed, I'll squash it later.

lukash commented:
I think all issues are resolved and this is ready to merge (along with the tests).

j-mracek commented:
@rh-atomic-bot r+

j-mracek commented:
@rh-atomic-bot retry

lukash commented:
@rh-atomic-bot retry

j-mracek commented:
@rh-atomic-bot retry

pkratoch commented:
The tests passed locally, merging.
